bigtable-hbase-dataflow/src/main/java/com/google/cloud/bigtable/dataflow/CloudBigtableIO.java

```diff
@@ -0,0 +1,440 @@
+import org.apache.hadoop.hbase.client.Result;
+import org.apache.hadoop.hbase.client.ResultScanner;
+import org.apache.hadoop.hbase.client.RetriesExhaustedWithDetailsException;
+import org.apache.hadoop.hbase.client.Scan;
+import org.apache.hadoop.hbase.client.Table;
+import org.apache.hadoop.hbase.util.Pair;
+
+import com.google.cloud.bigtable.hbase1_0.BigtableConnection;
+import com.google.cloud.dataflow.sdk.Pipeline;
+import com.google.cloud.dataflow.sdk.coders.Coder;
+import com.google.cloud.dataflow.sdk.coders.CoderRegistry;
+import com.google.cloud.dataflow.sdk.coders.VoidCoder;
+import com.google.cloud.dataflow.sdk.io.BoundedSource;
+import com.google.cloud.dataflow.sdk.options.PipelineOptions;
+import com.google.cloud.dataflow.sdk.transforms.DoFn;
+import com.google.cloud.dataflow.sdk.transforms.PTransform;
+import com.google.cloud.dataflow.sdk.transforms.ParDo;
+import com.google.cloud.dataflow.sdk.values.PCollection;
+import com.google.cloud.dataflow.sdk.values.PDone;
+
+/**
+ * CloudBigtableIO is the gateway for Dataflow / Cloud Bigtable setup, {@link Sink}s and @{link
+ * {@link Source}s. NOTE: Sinks are ready, and Sources are coming soon.
+ */
+
+public class CloudBigtableIO {
+  /**
+   * A {@link BoundedSource} for a Bigtable {@link Table}, which is potentially filtered by a
+   * {@link Scan}.
+   */
+  public static class Source extends BoundedSource<Result> {
+    private static final long serialVersionUID = -5580115943635114126L;
+
+    private final CloudBigtableScanConfiguration configuration;
+
+    /**
+     * A {@link BoundedSource} for a Bigtable {@link Table} with a start/stop key range, along with
+     * a potential filter via a {@link Scan}.
+     */
+    private static class SourceWithKeys extends BoundedSource<Result> {
+      private static final long serialVersionUID = 2561066007121040429L;
+
+      private final CloudBigtableScanConfiguration configuration;
+      private final byte[] startKey;
+      private final byte[] stopKey;
+
+      public SourceWithKeys(
+          CloudBigtableScanConfiguration configuration, byte[] startKey, byte[] stopKey) {
+        this.configuration = configuration;
+        this.startKey = startKey;
+        this.stopKey = stopKey;
+      }
+
+      @Override
+      public long getEstimatedSizeBytes(PipelineOptions options) throws Exception {
+        return 0;
+      }
+
+      @Override
+      public boolean producesSortedKeys(PipelineOptions options) throws Exception {
+        return true;
+      }
+
+      @Override
+      public Coder<Result> getDefaultOutputCoder() {
+        return new HBaseResultCoder();
+      }
+
+      @Override
+      public List<? extends BoundedSource<Result>> splitIntoBundles(long arg0, PipelineOptions arg1)
+          throws Exception {
+        return Collections.singletonList(this);
+      }
+
+      @Override
+      public void validate() {
+        checkNotNullOrEmpty(configuration.getProjectId(), "project");
```

Comment:
I think it's good to implement validation for each sub-source, in case the keys get deleted for example. We'll be validating each sub-source before reading it. But we can add a TODO and proceed if this is not possible now.